WeatherApp/Services/Persistance/CoreDataManager.swift

```diff
@@ -0,0 +1,145 @@
+//
+//  CoreDataManager.swift
+//  WeatherApp
+//
+//  Created by Lucija Balja on 24/08/2020.
+//  Copyright © 2020 Lucija Balja. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+class CoreDataManager {
+    
+    public typealias CoreDataManagerCompletion = () -> ()
+    private let completion: CoreDataManagerCompletion
+    
+    private let modelName: String
+    
+    init(modelName: String,  completion: @escaping CoreDataManagerCompletion) {
+        self.modelName = modelName
+        self.completion = completion
+        
+        setupCoreDataStack()
+    }
+    
+    //MARK: - Core Sata Stack Setup
+    
+    private lazy var managedObjectModel: NSManagedObjectModel = {
+        // Fetch Model URL
+        guard let modelURL = Bundle.main.url(forResource: self.modelName, withExtension: "momd") else {
+            fatalError("Unable to Find Data Model")
+        }
+        
+        // Initialize Managed Object Model
+        guard let managedObjectModel = NSManagedObjectModel(contentsOf: modelURL) else {
+            fatalError("Unable to Load Data Model")
+        }
+        
+        return managedObjectModel
+    }()
+    
+    private func setupCoreDataStack() {
+        // Fetch Persistent Store Coordinator
+        guard let persistentStoreCoordinator = mainManagedObjectContext.persistentStoreCoordinator else {
+            fatalError("Unable to Set Up Core Data Stack")
+        }
+        
+        DispatchQueue.global().async {
+            // Add Persistent Store
+            self.addPersistentStore(to: persistentStoreCoordinator)
+            
+            // Invoke Completion On Main Queue
+            DispatchQueue.main.async { self.completion() }
```

Comment:
Do not use `DispatchQueue.main.async` here. completion block should use that if necessary.